sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,958 @@
+			Parameters: asn1.RawValue{Tag: 5 /* ASN.1 NULL */},
+		},
+		nameHash,
+		issuerKeyHash,
+		serialNumber,
+	})
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(encodedCertID), nil
+}
+
+func (c *Client) checkOCSPResponseCache(encodedCertID *certIDKey, subject, issuer *x509.Certificate) (*ocspStatus, error) {
+	c.ocspResponseCacheLock.RLock()
+	gotValueFromCache := c.ocspResponseCache[*encodedCertID]
+	c.ocspResponseCacheLock.RUnlock()
+
+	status, err := c.extractOCSPCacheResponseValue(gotValueFromCache, subject, issuer)
+	if err != nil {
+		return nil, err
+	}
+	if !isValidOCSPStatus(status.code) {
+		c.deleteOCSPCache(encodedCertID)
+	}
+	return status, err
+}
+
+func (c *Client) deleteOCSPCache(encodedCertID *certIDKey) {
+	c.ocspResponseCacheLock.Lock()
+	delete(c.ocspResponseCache, *encodedCertID)
+	c.cacheUpdated = true
+	c.ocspResponseCacheLock.Unlock()
+}
+
+func validateOCSP(ocspRes *ocsp.Response) (*ocspStatus, error) {
+	curTime := time.Now()
+
+	if ocspRes == nil {
+		return nil, errors.New("OCSP Response is nil")
+	}
+	if !isInValidityRange(curTime, ocspRes.ThisUpdate, ocspRes.NextUpdate) {
+		return &ocspStatus{
+			code: ocspInvalidValidity,
+			err:  fmt.Errorf("invalid validity: producedAt: %v, thisUpdate: %v, nextUpdate: %v", ocspRes.ProducedAt, ocspRes.ThisUpdate, ocspRes.NextUpdate),
+		}, nil
+	}
+	return returnOCSPStatus(ocspRes), nil
+}
+
+func returnOCSPStatus(ocspRes *ocsp.Response) *ocspStatus {
+	switch ocspRes.Status {
+	case ocsp.Good:
+		return &ocspStatus{
+			code: ocspStatusGood,
+			err:  nil,
+		}
+	case ocsp.Revoked:
+		return &ocspStatus{
+			code: ocspStatusRevoked,
+		}
+	case ocsp.Unknown:
+		return &ocspStatus{
+			code: ocspStatusUnknown,
+			err:  errors.New("OCSP status unknown."),
+		}
+	default:
+		return &ocspStatus{
+			code: ocspStatusOthers,
+			err:  fmt.Errorf("OCSP others. %v", ocspRes.Status),
+		}
+	}
+}
+
+// retryOCSP is the second level of retry method if the returned contents are corrupted. It often happens with OCSP
+// serer and retry helps.
+func (c *Client) retryOCSP(
+	ctx context.Context,
+	client clientInterface,
+	req requestFunc,
+	ocspHost *url.URL,
+	headers map[string]string,
+	reqBody []byte,
+	issuer *x509.Certificate) (
+	ocspRes *ocsp.Response,
+	ocspResBytes []byte,
+	ocspS *ocspStatus, err error) {
+
+	request, err := req("POST", ocspHost.String(), bytes.NewBuffer(reqBody))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if request != nil {
+		request = request.WithContext(ctx)
+		for k, v := range headers {
+			request.Header[k] = append(request.Header[k], v)
+		}
+	}
+	res, err := client.Do(request)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	defer res.Body.Close()
+	c.Logger().Debug("StatusCode from OCSP Server:", "statusCode", res.StatusCode)
+	if res.StatusCode != http.StatusOK {
+		return nil, nil, nil, fmt.Errorf("HTTP code is not OK. %v: %v", res.StatusCode, res.Status)
+	}
+	c.Logger().Debug("reading contents")
```

Comment:
Not sure if this Debug statement is worth keeping?

Reply:
Yep, these were all Info statements in the original, very chatty.